src/surrogate_heat_driver.cpp

```diff
@@ -87,40 +180,249 @@ void SurrogateHeatDriver::generate_arrays()
+  fluid_density_ = xt::empty<double>({n_pins_, n_axial_});
+}
+
+double SurrogateHeatDriver::rod_axial_node_power(const int pin, const int axial) const
+{
+  Expects(axial < n_axial_);
+
+  double power = 0.0;
+  double dz = z_(axial + 1) - z_(axial);
+
+  for (gsl::index i = 0; i < n_rings(); ++i)
+    power += source_(pin, axial, i) * solid_areas_(i) * dz;
+
+  return power;
 }
 
 void SurrogateHeatDriver::solve_step()
 {
-  std::cout << "Solving heat equation...\n";
-  // NuScale inlet temperature
-  double T_co = 523.15;
+  solve_fluid();
+  solve_heat();
+}
 
-  // Set initial temperature
-  for (auto& T : temperature_) {
-    T = T_co;
+void SurrogateHeatDriver::solve_fluid()
+{
+  // determine the power deposition in each channel; the target applications will always
+  // be steady-state or pseudo-steady-state cases with no axial conduction such that
+  // the power deposition in each channel is independent of a convective heat transfer
+  // coefficient and only depends on the rod power at that axial elevation. The channel
+  // powers are indexed by channel ID, axial ID
+  xt::xtensor<double, 2> channel_powers({n_channels_, n_axial_}, 0.0);
+  for (int i = 0; i < n_channels_; ++i) {
+    for (int j = 0; j < n_axial_; ++j) {
+      for (const auto& rod : channels_[i].rod_ids_)
+        channel_powers(i, j) += 0.25 * rod_axial_node_power(rod, j);
+    }
   }
 
+  // initial guesses for the fluid solution are uniform temperature (set to the inlet
+  // temperature) and uniform pressure  (set to the outlet pressure). These solution
+  // fields are defined on channel axial faces. The units used throughout this section
+  // are h (kJ/kg), P (MPa), u (m/s), rho (kg/m^3). Unit conversions are performed as
+  // necessary on the converged results before being used in the Monte Carlo solver.
+  // Enthalpy here requires a factor of 1e-3 to convert from J/kg to kJ/kg.
+  xt::xtensor<double, 2> h({n_channels_, n_axial_ + 1}, iapws::h1(pressure_bc_, inlet_temperature_) * 1e-3);
+  xt::xtensor<double, 2> p({n_channels_, n_axial_ + 1}, pressure_bc_);
+
+  // for certain verbosity settings, we will need to save the velocity solutions
+  xt::xtensor<double, 2> u = xt::zeros<double>({n_channels_, n_axial_ + 1});
+
+  bool converged = false;
+  for (gsl::index iter = 0; iter < max_subchannel_its_; ++iter) {
+    // save the previous solution
+    xt::xtensor<double, 2> h_old = h;
+    xt::xtensor<double, 2> p_old = p;
+
+    // solve each channel independently
+    for (gsl::index chan = 0; chan < n_channels_; ++chan) {
+      const auto& c = channels_[chan];
+
+      // solve for enthalpy by simple energy balance q = mdot * dh by marching from inlet;
+      // divide term on RHS by 1e3 to convert from J/kg to kJ/kg
+      h(chan, 0) = iapws::h1(p(chan, 0), inlet_temperature_);
+      for (gsl::index axial = 0; axial < n_axial_; ++axial)
+        h(chan, axial + 1) = h(chan, axial) + channel_powers(chan, axial) / channel_flowrates_(chan) / 1e3;
```

Comment:
```suggestion
        h(chan, axial + 1) = h(chan, axial) + 1e-3 * channel_powers(chan, axial) / channel_flowrates_(chan);
```
Multiplication is more efficient than division on most architectures.